Add support for mounting folders from iOS

NoahPeeters opened the pull request:
Adds a button to the bar to mount any folder from a FileProvider into ish as described in #230.

Not all FileProvider support giving read/write access to a folder. Those aren't available.

Working FileProviders:
- On My iPad/iPhone
- iCloud Drive
- Working Copy
- SMB servers

FileProviders which do not work
- GoogleDrive
- ish itself
- OneDrive

tbodt commented:
Thanks! There are a few things missing, such as file coordination.

tbodt commented on app/ExternalFolder.m:
```diff
@@ -0,0 +1,56 @@
+    return fs;
+}
+
+const struct fs_ops *externalfs = nil;
```
Can you create this statically to avoid having to call something to create it?

tbodt commented on app/ExternalFolder.m:
```diff
@@ -0,0 +1,56 @@
+    NSURL *url = urlForMount(mount);
+    if ([url startAccessingSecurityScopedResource] == NO) {
+        CFBridgingRelease(mount->data);
+        return errno_map();
```
startAccessingSecurityScopedResource doesn't set errno so this will return garbage

tbodt commented on fs/mount.c:
```diff
@@ -37,15 +37,15 @@ void mount_release(struct mount *mount) {
 }
 
-int do_mount(const struct fs_ops *fs, const char *source, const char *point, int flags) {
+int do_mount_with_data(const struct fs_ops *fs, const char *source, const char *point, int flags, void *data) {
```
You're supposed to set mount->data after calling this, there shouldn't need to be a new function

NoahPeeters replied:
My problem was (and still is) that do_mount calls `fs->mount` and at that point `startAccessingSecurityScopedResource` must be called. For that, I need the NSURL object.

NoahPeeters replied:
My problem was (and still is) that `do_mount` calls `fs->mount` and at that point `startAccessingSecurityScopedResource` must be called. For that, the `NSURL` object is required.

tbodt commented on app/TerminalViewController.m:
```diff
@@ -234,6 +238,48 @@ - (IBAction)pressArrow:(ArrowBarButton *)sender {
+    [self presentViewController:picker animated:true completion:nil];
+}
+
+-(void)displayMountError {
```
Should include some information on what the error is

tbodt commented:
Also I think a better idea for the UI is to open the UIDocumentPickerViewController code in the mount function. That way you type something like `mount -t ios website /mnt/website` and it pops up a file picker.

NoahPeeters commented:
I've updated the pull request. Now, new directories can be mounted with the `mount -t ...` command.
I also included the `NSFileCoordinator` but it is currently not used for opening files as I ran into some problems. One simple example is creating a fifo with mkfifo. If one now tries to read it with cat and then pipe data into the fifo with echo, the echo command blocks because it requires write access and cat still has read access to the file.

tbodt commented on app/iOSFS.m:
```diff
@@ -0,0 +1,320 @@
+}
+
+static int iosfs_mount_data(char source[MAX_PATH], char point[MAX_PATH], dword_t *flags, void **data) {
+    pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
```
Can this be moved into iosfs_mount? I think that would eliminate the need for a new do_mount_with_data.

tbodt commented:
File coordination was clearly not designed for fifos. I think it's reasonable to check what type the file is, and only use file coordination for regular files where read/write can't block.

videoMonkey commented:
Would love to see this happen, is any progress being made? I wish my programming chops were at the level to be able to help.

NoahPeeters commented:
I had to do a lot for university lately. I will definitely continue working on this. Probably between Christmas and New Year or so. I already looked into it and it is not that hard to implement it.

videoMonkey commented:
Hey! Thanks!

—Jamie

> On Dec 12, 2019, at 10:28, Noah Peeters <notifications@github.com> wrote:
> 
> ﻿
> I had to do a lot for university lately. I will definitely continue working on this. Probably between Christmas and New Year or so. I already looked into it and it is not that hard to implement it.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.

NoahPeeters commented:
I continued working on this. I added file coordination for regular files and removed the `iosfs_mount_data` function by calling the ui functions in `iosfs_mount`.

I did a lot of testing with `git` in a mounted folder from Working Copy. I noticed that `git status` can be quit slow even with mid-size repositories because git will check every file in the repository and each of the coordinator calls slows down the process. That's why I added a second filesystem: `ios-unsafe`. That one will simply make sure that the correct permissions for the security scope are requested but the filecoordinators are not used at all. In case of Working Copy, this works very well but there are probably apps which can have a problem with that (that's why I called in `ios-UNSAFE`).

Therefore, this pull request now adds the following two commands:

```sh
mount -t ios website /mnt/website
mount -t ios-unsafe website /mnt/website
```

Finally, someone with more experience in async programming in objective-c should have a look at the usage of mutexes in `iosfs_open`. It does work but I do not know if there is a better/more elegant way of doing this.

tbodt commented:
Nice! Will do a detailed review in a bit.

I'm torn on whether ios-unsafe is a good idea. For the specific case of Working Copy, it makes more sense to just do git operations directly in the Working Copy app.

tbodt commented on app/SceneDelegate.h:
```diff
@@ -6,9 +6,12 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+extern TerminalViewController *currentTerminalViewController;
```
I don't really like this way of getting the view controller. My other ideas are way more annoying to implement though.

tbodt commented on app/iOSFS.m:
```diff
@@ -0,0 +1,449 @@
+@interface DirectoryPickerDelegate : NSObject <UIDocumentPickerDelegate>
+
+@property NSURL *url;
+@property pthread_mutex_t *mutex;
```
Locks aren't really intended for waiting for something to happen. Should use a condition variable. cond_t from util/sync.h is preferred because it allows the process to be interrupted by a signal, with e.g. ctrl-c.

tbodt commented on app/iOSFS.m:
```diff
@@ -0,0 +1,449 @@
+    return combine_error(error, err);
+}
+
+static int iosfs_symlink(struct mount *mount, const char *target, const char *link) {
```
There's a lot of repeated code in these functions. How much can be factored out into a new function that takes a block?

tbodt commented:
@NoahPeeters Do you plan to work on this soon? If not, I'm willing to pick it up and get it merged.

tbodt commented:
@NoahPeeters Ping

ThisIsQasim commented:
@NoahPeeters really appreciate the work you have done so far and if you could get this to completion it would really supercharge my device and open it up for endless possibilities.

MrAlexWeber commented:
+1 for this feature, replying to bump and watch

tbodt commented:
Gonna take this over

tbodt commented:
Merged in dbd4100.
